### ISSUE TITLE
perf(fri): close the 2.52× FRI-stage gap vs native pil2 (the fold)

### ISSUE THREAD

**Issue by erkhesbold2005-blip** [fractalyze/zisk-zorch#65]
## Goal

Close the FRI prover stage's **2.52× gap** vs the native pil2-stark CUDA prover on the same hardware. The gap is almost entirely the **fold**; the per-layer Merkle commit is already at parity.

## Why

Baselining the per-stage prover on an RTX 5090 (real Main-air dims, N=2²² → N_ext=2²³) leaves FRI as the largest remaining gap after commit (1.04×, done) and extend (0.58×, #58): FRI is where zorch actually trails pil2.

**Measured (RTX 5090, schedule `[23,20,17,14,11,8,5]` = fold 3 bits/layer, Merkle arity 4, query phase excluded):**

| component | zisk-zorch | native pil2 | ratio |
|---|---|---|---|
| **fold** | **14.75 ms** | **3.29 ms** | **4.48×** |
| merkle | ~5.09 ms | 4.56 ms | ~1.12× (parity) |
| **total** | **19.84 ms** | **7.86 ms** | **2.52×** |

Per-layer, the first fold (2²³→2²⁰) dominates both sides and is where we lose: **12.67 ms vs 2.79 ms**; later layers decay geometrically. The Merkle portion uses the same Poseidon2 tree as the trace commit (1.04×), so it is not the gap.

## Root cause — per-group Lagrange fold vs pil2's shared-twiddle INTT

- **Ours** (`zorch.coding.reed_solomon.fri_fold_k_values`, driven by `zisk_zorch/fri/fold.py`): a **per-group Lagrange interpolation**. For step 1 the coset domain is a `(2²⁰ × 8)` array — a *different* 8-point set per group — so an O(n²) Lagrange basis is recomputed **independently for all 2²⁰ groups**. This is compute-bound: 12.67 ms to move a 200 MB codeword = ~15 GB/s effective, far below the ~1.7 TB/s the GPU can stream, so it is not bandwidth-limited. `fold.py` chose Lagrange deliberately ("free of any NTT root-order convention").
- **pil2** (`starks_gpu.cu` `fold` → `intt_tinny`): a size-`n_x` **INTT with twiddles shared across all groups** (precomputed once per block in shared memory) + coset rescale + Horner-at-challenge. O(n log n), and it exploits that every group folds on the *same relative root* — the structure our per-group Lagrange discards.

## Proposed levers (either, or both)

1. **Shared-basis / INTT-style k-ary fold.** Restructure the fold so the fold coefficients are shared across groups instead of a per-group Lagrange basis — i.e. an in-thread INTT + coset rescale + Horner like pil2, or a factored Lagrange where the per-group work is only the coset shift. This is a change to the **zorch** fold primitive (`fri_fold_k_values` / a new shared-twiddle variant); `zisk_zorch/fri/fold.py` then calls it. **Cross-repo — the primitive is in zorch, the stage is here** (cf. #58's xla dependency).
2. **Factor-8 fold as 3× k=2 butterflies.** zorch's own note (`reed_solomon.py:544-547`) says the k=2 conjugate-pair butterfly is cheaper than its Lagrange form; a factor-8 layer = three successive halving butterflies could beat one k=8 Lagrange. Trades kernel count for cheaper per-element work; measure vs lever 1.

Secondary: the cubic dtype SoA layout (fractalyze/xla_fork#258) would raise the fold's memory ceiling, but the fold is *compute*-bound here, so the algorithmic fix above is primary.

## End criteria

- [x] FRI fold drops from 14.75 ms toward pil2's 3.29 ms at N_ext=2²³ (target: within ~1.2× of native).
- [x] FRI stage total approaches pil2's 7.86 ms (from 19.84 ms).
- [x] `fri/fold_test.py` + `fri/prover_test.py` still byte-match (the fold is byte-identical to pil2's INTT-rescale by construction — a shared-twiddle rewrite must preserve that).

## Repro / harness

- Ours: `bench_inner_proof --stages=fri --n_bits=22 --blowup_bits=1 --arity=4` (gpu-venv, pristine pinned plugin). Fold-only decomposition: time `zisk_zorch/fri/fold.py::fold` jitted per step (`/tmp/claude-1006/gsumbench/fri_fold_only.py`).
- pil2 proxy: `/tmp/claude-1006/fri_bench` (lifts `fold`/`intt_tinny`/`transposeFRI` + `Poseidon2GoldilocksGPU<16>::merkletree`; source `/tmp/claude-1006/gsumbench/fri_bench.cu`, build `/tmp/claude-1006/build_run_fri.sh`).
- ⚠️ Do **not** `jax.profiler.trace` the FRI prove at 2²³ — it leaks host RAM and spikes CPU load; decompose by direct sub-op timing instead.

Note: the LogUp grand-sum optimizations (#64, prime-ir #398) do **not** touch the FRI code path — this gap is independent of them.

**Comment by erkhesbold2005-blip** [comment]
### Session 2026-07-12 progress — started the fold optimization

**Done this session:** baselined (see issue body: fold 14.75ms vs pil2 3.29ms = 4.48×, compute-bound on the per-group Lagrange), then prototyped the first candidate fix.

**Key finding — the "share the Lagrange denominators" shortcut is a near dead-end (1.03×).** The current `fold.py` passes a *different* point-set per group (`domain[g] = scale_g·relative_points`), so `_lagrange_denominators(domain[g])` looks recomputed per group. Using the Lagrange scale-invariance identity `L_k^{s·p}(β)=L_k^{p}(β/s)`, I restructured the fold to pass **shared relative points + a per-group β/scale** (so the denominators become a vmap-constant XLA hoists once), which also shrinks the coset domain from a `(2²⁰,n_x)` memory array to `n_x` registers + a per-group scalar.

Result (byte-match=True both sizes, `/tmp/claude-1006/gsumbench/fri_fold_opt.py`):
- `2^21→2^18`: old 3.176ms → new 3.050ms = **1.04×**
- `2^23→2^20`: old 12.636ms → new 12.212ms = **1.03×**

So sharing the denominators does **not** move it — XLA already CSE'd them, and the real cost is the **O(n_x²) per-group numerators × 2²⁰ groups** (`compute_lagrange_basis`'s `jnp.prod(where(mask,...))`), plus the vmap-over-2²⁰ structure. This shortcut is byte-correct but not worth landing on its own.

**Next session — the real lever is the INTT-style fold (pil2's `intt_tinny`).** Reimplement the k-ary fold as a size-`n_x` INTT (O(n_x log n_x), twiddles **shared** across all groups in registers/smem) + coset rescale + Horner-eval at β — exactly what pil2's `fold` kernel does. For n_x=8 that's ~24 butterfly + ~16 rescale/Horner ops vs Lagrange's ~64+ numerator ops, and the shared twiddles are the structural win the Lagrange throws away.

- pil2 reference: `intt_tinny` at `pil2-stark/src/starkpil/starks_gpu.cu:643`, the `fold` kernel `:689-771`, wrapper `fold_inplace:773-800` (shows omega_inv/invShiftPow/invW setup).
- Our fold driver: `zisk_zorch/fri/fold.py::fold` (builds `_coset_domain`, `jax.vmap(fold_group)` over groups). The fold primitive `fri_fold_k_values` is in **zorch** `zorch/coding/reed_solomon.py:538`; the O(n²) basis is `zorch/poly/univariate.py::compute_lagrange_basis:54`. So the INTT fold likely lands as a new **zorch** k-ary-INTT-fold primitive (or an `is_dif`-style INTT eval) that `fold.py` calls.

**Investigation recipe (copy-paste):**
```
# env: this box needs CUDA_VISIBLE_DEVICES=0 (device enum wedges otherwise);
# XLA_PYTHON_CLIENT_PREALLOCATE=false; NEVER jax.profiler.trace at 2^23 (leaks RAM, wedges CUDA).
GPUVENV=/tmp/claude-1006/-home-eric-Workspace/9e16cf9b-d34f-4ce4-97e0-1dcf24775e53/scratchpad/gpu-venv
# 1) baseline our fold per step (dominant step-1 = 2^23->2^20):
CUDA_VISIBLE_DEVICES=0 PYTHONPATH=/home/eric/Workspace/zisk-zorch:/home/eric/Workspace/envs/zisk-zorch/zorch \
  $GPUVENV/bin/python /tmp/claude-1006/gsumbench/fri_fold_only.py
# 2) prototype the INTT fold; byte-match vs fold_old + time, mirroring fri_fold_opt.py's harness.
# 3) pil2 target: /tmp/claude-1006/fri_bench (fold_total 3.29ms) — already built.
```

**Decision tree after the INTT prototype:**
- byte-match True AND ≥2× faster → implement the INTT k-ary fold in **zorch** (new primitive, seam-conformant per `zorch` docs/conventions), rewire `fold.py`, run `zisk_zorch/fri/fold_test.py` (golden byte-match) + `fri/prover_test.py`, then PR.
- byte-match True but NOT faster → the bottleneck is the **vmap-over-2²⁰ / memory layout / the `.T` transpose in `group = pol.reshape(n_x,cur_n).T`**, not the arithmetic; profile the fold kernel's HLO (dump, don't `jax.profiler` — use `.lower().compile().as_text()`) for a transpose/gather boundary and restructure the group access.
- byte-match False → the coset root-order / `is_inverse` twiddle convention differs from pil2; reconcile against `fold.py`'s `_coset_domain` (`_TWO_ADIC_ROOT`, `shift_eff`) — the fold is byte-identical "by IDFT uniqueness" per `fold.py`'s docstring, so the value must match however computed.

**Reference files:**
- Source to change: `/home/eric/Workspace/zisk-zorch/zisk_zorch/fri/fold.py`; `/home/eric/Workspace/envs/zisk-zorch/zorch/zorch/coding/reed_solomon.py` (`fri_fold_k_values:538`); `/home/eric/Workspace/envs/zisk-zorch/zorch/zorch/poly/univariate.py` (`compute_lagrange_basis:54`)
- Tests (byte-match): `zisk_zorch/fri/fold_test.py`, `fri/prover_test.py`
- pil2 ref + bench: `/tmp/claude-1006/pil2-proofman/pil2-stark/src/starkpil/starks_gpu.cu`, `/tmp/claude-1006/fri_bench` (`build_run_fri.sh`)
- Prototype harness: `/tmp/claude-1006/gsumbench/fri_fold_opt.py`, `fri_fold_only.py`

**Done criteria:** our fold drops from 14.75ms toward pil2's 3.29ms at N_ext=2²³ (target ≤~1.2×); FRI total 19.84→~9ms; `fold_test`/`prover_test` byte-match green; PR opened.

**Out of scope:** the FRI merkle (~parity, not the gap); the 3× k=2-butterfly alternative (changes the factor-8 commit structure → breaks the pil2 byte-match, rejected).

**Comment by erkhesbold2005-blip** [comment]
### Session 2026-07-12 10:53 progress — INTT fold implemented, all criteria met

**Done:** implemented the INTT-style shared-twiddle fold the last session pointed to, byte-match green, landed as a cross-repo change (committed locally, not pushed).

**Result (RTX 5090, schedule `[23,20,17,14,11,8,5]`):**

| | baseline (Lagrange) | new (INTT) | pil2 |
|---|---|---|---|
| fold total | 14.75 ms | **1.42 ms** | 3.29 ms |
| dominant step-1 (2²³→2²⁰) | 12.66 ms | **1.22 ms** | 2.79 ms |
| FRI stage total | 19.84 ms (2.52×) | **~6.5 ms (~0.83×)** | 7.86 ms |

**10.4× over baseline, and 2.3× below pil2's own fold** — past the ≤1.2×-of-native target. All three End-criteria boxes now ticked.

**Implementation (2 repos, tightly scoped — committed on branch `perf/fri-fold-intt` in each, no push):**
- **zorch** `zorch/coding/reed_solomon.py`: new agnostic primitive `fri_fold_k_coset` — bit-reverse + DIT butterfly INTT (twiddles **shared** across all groups), coset rescale (`coeffₘ ·= s⁻ᵐ`), Horner-at-β. Unrolled over the static factor `k`, coalesced over the batched group axis (O(k log k) vs Lagrange's O(k²)-per-group). Test added vs `fri_fold_k_values` on a coset, k∈{2,4,8}, batched.
- **zisk-zorch** `zisk_zorch/fri/fold.py`: prover `fold` calls `fri_fold_k_coset` with `(omega_inv, coset_inv)` built from the existing pil2-root coset setup; `verify_fold` stays Lagrange (per-query, byte-equal by interpolant uniqueness).

**Why byte-match holds:** the degree-<n_x interpolant through the coset points is unique, so its value at β is one field element however computed. Verified at n_x∈{2,8,32}, first/later steps, 2-step chains, and the full pil2 prover golden.

**Native `lax.ntt` rejected:** its canonical root is ω⁵ (empirically confirmed), needing a per-group reindex against pil2's coset root; the hand-unrolled butterfly on the caller's explicit ω sidesteps that. The zorch "no hand-rolled butterfly" rule targets the long-axis LDE NTT, not a batched size-k INTT — documented in the primitive.

**Tests green:** zisk `fold_test` (2), `prover_test` (1), `seam_test`, `verifier_test` (2); zorch `reed_solomon_test` (50, incl. the new coset test). zorch source-lint clean (agnostic + fusion).

**Next session:** cross-repo landing (issue anticipated this, cf. #58) — the zisk `fold.py` commit depends on the new zorch `fri_fold_k_coset`, so the zorch branch merges first/together. Both branches are local commits awaiting push/PR review. The zisk change is isolated to `fold.py`; that repo's working tree also carries 26 unrelated in-flight files (branch `feat/verify-trace-commit`) that the new branch was cut from — rebase onto `main` before a PR.

**Blocker:** none.

**Comment by erkhesbold2005-blip** [comment]
**Update — PRs open.** Pushed both branches and opened PRs:
- zisk-zorch **#70** (`fold.py` only, `Closes #65`) — https://github.com/fractalyze/zisk-zorch/pull/70
- zorch **#456** (new `fri_fold_k_coset` primitive + test) — https://github.com/fractalyze/zorch/pull/456

Cross-repo dependency: zorch #456 merges first (zisk #70's `fold.py` imports `fri_fold_k_coset`). The zisk commit was rebased clean onto `main` (the earlier working-tree base carried an unrelated plain-Goldilocks import change, now dropped).
